### ISSUE TITLE
UTF-8 in source code with Mono

### ISSUE THREAD

**Issue by weinholt** [leppie/IronScheme#31]
Hello @leppie! I have a bunch of source code where the files are encoded in UTF-8. I can't seem to get IronScheme to read them with the right encoding; they appear to be read as iso-8859-1:

```
# export LANG=C.UTF-8
# cat foo.scm
(import (rnrs))
(display "åäö\n")
# mono /usr/local/IronScheme/IronScheme.Console-v4.exe foo.scm
IronScheme 1.0.110-6a2cd45 github.com/leppie/IronScheme © 2007-2016 Llewellyn Pritchard (.NET 4.0 64-bit)
Ã¥Ã¤Ã¶
```

This is with Mono 5.18.0.240+dfsg-3 on Debian 10.0 (amd64). The exact environment is available in Docker as weinholt/ironscheme:latest.

Any suggestions?

**Comment by leppie** [comment]
Hi @weinholt 

I can replicate and this happens on Windows/MS version too. While I havent looked too deeply into it, it seems to guess the content wrong. It works perfectly if you save with a BOM. 

Will look into it to see if it is a regression of sorts.

Thanks

**Comment by leppie** [comment]
@weinholt Thanks, it will default to utf8 reader now on source/scheme files loaded. Release created

**Comment by weinholt** [comment]
@leppie Works beautifully, thanks!

Found a related problem, will open a new issue.

**Comment by weinholt** [comment]
@leppie Actually, I'll continue here, it seems very related. UTF-8 in files started working. However, something is wonky with UTF-8 in symbols. (It works in the REPL, and fails in files, which is why I continue on the same issue).

```scheme
# cat > foo.scm
(import (rnrs (6)))
(write "ö")
(newline)
# mono /usr/local/IronScheme/IronScheme.Console-v4.exe foo.scm
IronScheme 1.0.115-c56e079 github.com/leppie/IronScheme © 2007-2016 Llewellyn Pritchard (.NET 4.0 64-bit)
"ö"
# cat > bar.scm
(import (rnrs (6)))
(write 'ö)
(newline)
# mono /usr/local/IronScheme/IronScheme.Console-v4.exe bar.scm
IronScheme 1.0.115-c56e079 github.com/leppie/IronScheme © 2007-2016 Llewellyn Pritchard (.NET 4.0 64-bit)
&lexical
&message: "Bad UTF8 character: 182"
&irritants: ("(standard-input-port)")

# mono /usr/local/IronScheme/IronScheme.Console-v4.exe 
IronScheme 1.0.115-c56e079 github.com/leppie/IronScheme © 2007-2016 Llewellyn Pritchard (.NET 4.0 64-bit)
> (write 'ö)
ö>     
```

So it says "Bad UTF8 character: 182". It complains about the second byte: `(string->utf8 "ö")` = `#vu8(195 182)`.

**Comment by leppie** [comment]
The fix contradicts what I did in 2e4c3ed4

**Comment by weinholt** [comment]
I hope we will not need a battle royale to decide between latin-1 and utf-8.

The UTF-8 decoder in `TextBuff.Read` lacks checks (surrogate pairs, overlong encodings) and emoji support (🦆!), but the two-character case at least looks like it should work.

Don't know if it helps, but `(write ' ö)`, with a space after the quote, works as it should.

**Comment by leppie** [comment]
Smallest repro: `(ö)` 

Yes it is a very specific case.

**Comment by leppie** [comment]
Testing a fix now :D

**Comment by leppie** [comment]
Fixed in e8d1612 . Release coming soon.

**Comment by weinholt** [comment]
That seems to have done it. Thanks!
